### PR TITLE
prevent invalid scope when parsing bodies containing erroneous decls

### PR DESCRIPTION
**Pull request by Basile-z**
Required to fix ...  https://github.com/dlang-community/DCD/issues/600

### CODE REVIEW

**Comment by Basile-z** [comment]
this breaks the arbitrary lookups required by the language.

**Comment by WebFreak001** [comment]
if `parseDeclarationsAndStatements` can't return null anymore, I think the following classes/members would never be null anymore:
- `CaseStatement.declarationsAndStatement`
- `CaseRangeStatement.declarationsAndStatement`
- `DefaultStatement.declarationsAndStatement`

`BlockStatement.declarationsAndStatement` before was null with an error in the content, now it is really only null if it's empty (`}` following the `{`)

Because of this PR I also found the following potential null-reference-errors which would be fixed by this:

formatter.d:616 (`void format(const CaseRangeStatement caseRangeStatement)`) and formatter.d:638 (`void format(const CaseStatement caseStatement)`) both call `formatCaseDecls` without null check of `declarationsAndStatements` and the method itself just accesses members.

formatter.d:1036 (`void format(const DefaultStatement defaultStatement)`) directly accesses the member without null check

nvm stopping checking this out if it's deleted then

**Comment by Basile-z** [comment]
Yeah the formatter bugs are not a surprise, see #346.

**Comment by Basile-z** [comment]
@WebFreak001, I have maybe a working fix coming.

**Comment by Basile-z** [comment]
Okay it seems to work and doesn't break dparse test suite. This time the approach is more pragmatic. The case that breaks DCD, i.e a syntax error, is exactly detected.

File: src/dparse/parser.d [hunk 1]
```diff
@@ -2371,13 +2371,29 @@ class Parser
...
                 allocator.rollback(c);
+
+                // detect the pattern ".}" for DCD. This is what happens when
+                // located at the end of a well delimited body/scope and requesting
+                // completion. This is also a case where it's sure sure that
+                // there's no ambiguity, even if it happens during a lookup:
+                // it's not a decl, it's not a statement, it's an error.
+                if (currentIs(tok!"}") && index > 0 && previous == tok!".")
+                {
+                    advance();
```
Comment (Basile-z) [inline src/dparse/parser.d]: must not advance

**Comment by Basile-z** [comment]
DCD test suite work too now.

**Pull request merged**